src/quarot/hadamard_utils.py

```diff
@@ -0,0 +1,168 @@
+        K = 60
+        hadK = get_had60().T if transpose else get_had60()
+    elif n % 52 == 0:  # llama-1-13b 1x hidden
+        assert is_pow2(n // 52)
+        K = 52
+        hadK = get_had52().T if transpose else get_had52()
+    elif n % 36 == 0:
+        assert is_pow2(n // 36)
+        K = 36
+        hadK = get_had36().T if transpose else get_had36()
+    elif n % 28 == 0:
+        assert is_pow2(n // 28)
+        K = 28
+        hadK = get_had28().T if transpose else get_had28()
+    elif n % 40 == 0:
+        assert is_pow2(n // 40)
+        K = 40
+        hadK = get_had40().T if transpose else get_had40()
+    elif n % 20 == 0:
+        assert is_pow2(n // 20)
+        K = 20
+        hadK = get_had20().T if transpose else get_had20()
+    elif n % 12 == 0:
+        assert is_pow2(n // 12)
+        K = 12
+        hadK = get_had12().T if transpose else get_had12()
+    else:
+        assert is_pow2(n)
+        K = 1
+
+    return hadK, K
+
+
+def matmul_hadU(X, transpose=False):
+    n = X.shape[-1]
+    hadK, K = get_hadK(n, transpose)
+    input = X.clone().view(-1, n, 1)
+    output = input.clone()
+    while input.shape[1] > K:
+        input = input.view(input.shape[0], input.shape[1] // 2, 2, input.shape[2])
+        output = output.view(input.shape)
+        output[:, :, 0, :] = input[:, :, 0, :] + input[:, :, 1, :]
+        output[:, :, 1, :] = input[:, :, 0, :] - input[:, :, 1, :]
+        output = output.view(input.shape[0], input.shape[1], -1)
+        (input, output) = (output, input)
+    del output
+
+    if K > 1:
+        # Do not explicitly repeat - OOM
+        # input = torch.bmm(
+        #     hadK.repeat(len(input), 1, 1).to(input.device).to(input.dtype), input)
+        # Use bcast instead
+        input = hadK.view(1, K, K).to(input) @ input
+
+    return input.view(X.shape) / torch.tensor(n).sqrt()
+
+
+def matmul_hadUt(X):
+    return matmul_hadU(X, transpose=True)
+
+
+def random_hadamard_matrix(size: int, seed: int = 0):
+    """
+    Generates a random Hadamard matrix of size `size`. See https://cornell-relaxml.github.io/quip-sharp/ , Section "Randomized Hadamard Transformation" for details.
+    """
+    torch.manual_seed(seed)
+    Q = torch.randint(low=0, high=2, size=(size,)).to(torch.float64)
+    Q = Q * 2 - 1
+    Q = torch.diag(Q)
+    return matmul_hadU(Q)
+
+
+def matmul_hadU_cuda(X: torch.tensor, hadK: torch.tensor, K: int) -> torch.tensor:
+    n = X.shape[-1]
+    if K == 1:
+        return hadamard_transform(X.contiguous(), 1.0 / torch.tensor(n).sqrt())
+
+    input = X.view(-1, K, n // K)
+    input = hadamard_transform(input.contiguous(), 1.0 / torch.tensor(n).sqrt())
+    input = hadK.to(input.device).to(input.dtype) @ input
+    return input.reshape(X.shape)
+
+
+def matmul_hadUt_cuda(X, hadK, K):
+    return matmul_hadU_cuda(X, hadK, K, transpose=True)
+
+
+def apply_exact_had_to_linear(module: torch.nn.Linear, had_dim: int = -1, output: bool = False):
```

Comment:
this function name feels clunky. Can we get a docstring please?

Reply:
done